store-implementation/accumulo-store/src/main/java/uk/gov/gchq/gaffer/accumulostore/MiniAccumuloStore.java

```diff
@@ -0,0 +1,274 @@
+            baseDir = new File(targetDir, BASE_DIRECTORY + UUID.randomUUID());
+        } else {
+            baseDir = new File(FileUtils.getTempDirectory(), BASE_DIRECTORY + UUID.randomUUID());
+        }
+
+        try {
+            FileUtils.deleteDirectory(baseDir);
+            miniAccumuloConfig = new MiniAccumuloConfig(baseDir, ROOTPW);
+            miniAccumuloConfig.setInstanceName(getProperties().getInstance());
+            miniAccumuloCluster = new MiniAccumuloCluster(miniAccumuloConfig);
+            miniAccumuloCluster.start();
+        } catch (final IOException | InterruptedException e) {
+            throw new StoreException(e.getMessage(), e);
+        }
+
+        // Create the user specified in the properties (if not root)
+        // together with the specified password and give them all authorisations
+        try {
+            if (!getProperties().getUser().equalsIgnoreCase("root")) {
+                miniAccumuloCluster.getConnector("root", ROOTPW).securityOperations()
+                        .createLocalUser(getProperties().getUser(), new PasswordToken(getProperties().getPassword()));
+                miniAccumuloCluster.getConnector("root", ROOTPW).securityOperations()
+                        .grantSystemPermission(getProperties().getUser(), SystemPermission.CREATE_TABLE);
+                miniAccumuloCluster.getConnector("root", ROOTPW).securityOperations()
+                        .grantSystemPermission(getProperties().getUser(), SystemPermission.CREATE_NAMESPACE);
+            }
+            Authorizations auths = new Authorizations("public", "private", "publicVisibility", "privateVisibility", "vis1", "vis2");
+            miniAccumuloCluster.getConnector("root", ROOTPW).securityOperations()
+                    .changeUserAuthorizations(getProperties().getUser(), auths);
+        } catch (final AccumuloException | AccumuloSecurityException e) {
+            throw new StoreException(e.getMessage(), e);
+        }
+
+        // Create the new properties object to pass back, including connection items
+        AccumuloProperties accumuloProperties = (AccumuloProperties) properties.clone();
+        accumuloProperties.setInstance(miniAccumuloCluster.getInstanceName());
+        accumuloProperties.setZookeepers(miniAccumuloCluster.getZooKeepers());
+
+        return accumuloProperties;
+    }
+
+    public void tearDownTestDB() {
+        this.closeMiniAccumuloStore();
+    }
+
+    public MiniAccumuloCluster getMiniAccumuloCluster() {
+        return miniAccumuloCluster;
+    }
+
+    public Connector getMiniConnector() {
+        return miniConnector;
+    }
+
+    @Override
+    public void preInitialise(final String graphId, final Schema schema, final StoreProperties properties) throws StoreException {
+        setProperties(properties);
+        // This MiniAccumuloStore object may be being reused so we need to clean up some of the attributes
+        clearAllOperationalHandlers();
```

Comment:
I feel like if the store is being re-used then it shouldn't be using the PreInitialise() method. But perhaps my assumption is wrong.

Reply:
The store's re-use it at the heart of the changes being implemented.
'MockAccumuloStore's are/were always created afresh every time for every test in a test suite.
To do this with 'MiniAccumuloStore's would be a massive overhead so it was requested that tests will set up a mini Accumulo instance at the start of a test suite, use it for all the tests in the suite, and then tear it down at the end of the suite.
It was found that when some tests were migrated to this approach some tests were failing as pre-conditions set for one test would, naturally enough, persist in the subsequent tests. Given that JUnit will not guarantee any specific sequence for tests in a suite it is necessary to ensure the Accumulo instance is returned to its 'default' condition before the  instance is used for the next test.